api/src/main/java/me/lokka30/treasury/api/economy/currency/Currency.java

```diff
@@ -12,69 +12,224 @@
+     * @param currencyChar          a way to identify the currency via a character e.g. '$' for a dollar. can be null.
+     * @param roundedDigits         to which digit shall we round balances. can specify -1 for no rounding.
+     * @param conversionCoefficient coefficient at which this currency is going to be converted to other currencies.
+     * @param startingBalance       a function, giving a starting balance for the inputted player uuid, which can be null.
+     * @param balanceFormatter      a function, giving a human-readable format of a balance.
+     * @param names                 acceptable names of the currency. primary name shall be the first specified.
+     * @return new currency instance
+     */
+    @NotNull
+    public static Currency of(
+            @NotNull UUID currencyId,
+            @Nullable Character currencyChar,
+            int roundedDigits,
+            double conversionCoefficient,
+            @Nullable Function<UUID, Double> startingBalance,
+            @NotNull BiFunction<Double, Locale, String> balanceFormatter,
+            @NotNull String... names
+    ) {
+        Objects.requireNonNull(currencyId, "currencyId");
+        return new Currency(
+                currencyId,
+                currencyChar,
+                roundedDigits,
+                conversionCoefficient,
+                startingBalance,
+                balanceFormatter,
+                names
+        );
+    }
+
+    private final UUID currencyId;
+    private final Character currencyChar;
+    private final String[] names;
+    private final int roundedDigits;
+    private final double conversionCoefficient;
+    private final Function<UUID, Double> startingBalance;
+    private final BiFunction<Double, Locale, String> balanceFormatter;
+
+    private Currency(@Nullable UUID currencyId,
+                     @Nullable Character currencyChar,
+                     int roundedDigits,
+                     double conversionCoefficient,
+                     @Nullable Function<UUID, Double> startingBalance,
+                     @NotNull BiFunction<Double, Locale, String> balanceFormatter,
+                     @NotNull String... names
+    ) {
+        if (currencyId == null) {
+            this.currencyId = UUID.randomUUID();
+        } else {
+            this.currencyId = currencyId;
+        }
+        this.currencyChar = currencyChar;
+        this.roundedDigits = roundedDigits;
+        this.conversionCoefficient = conversionCoefficient;
+        if (startingBalance == null) {
+            this.startingBalance = ($) -> 0d;
+        } else {
+            this.startingBalance = startingBalance;
+        }
+        if (names.length < 1) {
+            throw new IllegalArgumentException("Empty array specified for currency name.");
+        }
+        this.balanceFormatter = Objects.requireNonNull(balanceFormatter, "balanceFormatter");
+        this.names = Objects.requireNonNull(names, "names");
```

Comment:
The throws and Objects.requireNonNull could be at the begining of the constructor

Reply:
done.